### PR TITLE
[BUGFIX beta] Optimization in .keys

### PR DESCRIPTION
**Pull request by marijaselakovic**
Hi,

I spotted two locations where for-in loops can be refactored to Object.keys and for loop (similarly as in https://github.com/emberjs/ember.js/commit/53b9f19a8fd737ed9dfede59316ec83825de44d2). 

Results of running tests suites in Chrome 42 with original version: 

> Tests completed in 58335 milliseconds.13213 assertions of 13213 passed, 0 failed.

And with modified version: 

> Tests completed in 54474 milliseconds.13213 assertions of 13213 passed, 0 failed.

If you use some performance benchmark it would be interesting to try.

### CODE REVIEW

**Comment by rwjblue** [comment]
Can you prefix this commit with `[BUGFIX beta]` so that we can pull that into the beta branch?

File: packages/ember-metal/lib/mixin.js [hunk 1]
```diff
@@ -647,11 +648,12 @@ MixinPrototype.keys = function() {
   var seen = {};
   var ret = [];
```
Comment (stefanpenner) [inline packages/ember-metal/lib/mixin.js]: This can be pre allocated

Comment (marijaselakovic) [inline packages/ember-metal/lib/mixin.js]: `var ret = Object.keys(keys)`, without for loop?

Comment (rwjblue) [inline packages/ember-metal/lib/mixin.js]: Yes, I believe that would work.

**Comment by stefanpenner** [comment]
Its worth noting, that for of vs Object.keys isn't always cut and dry. Lots of really annoying trade-offs.

**Comment by rwjblue** [comment]
This looks good.  Can you squash all the commits down into one?

File: packages/ember-metal/lib/mixin.js [hunk 2]
```diff
@@ -633,9 +633,10 @@ function _keys(ret, mixin, seen) {
   seen[guidFor(mixin)] = true;
 
   if (mixin.properties) {
-    var props = mixin.properties;
-    for (var key in props) {
-      if (props.hasOwnProperty(key)) { ret[key] = true; }
```
Comment (stefanpenner) [inline packages/ember-metal/lib/mixin.js]: @krisselden can we safely omit the own clause here ?

Comment (rwjblue) [inline packages/ember-metal/lib/mixin.js]: @stefanpenner - Doesn't `Object.keys` only return own properties?

Comment (stefanpenner) [inline packages/ember-metal/lib/mixin.js]: the semantically the output values would be the same (although sort order can differ depending on many things...), but that isn't why i asked.

If props is always a fast object and never has inherited properties (which maybe we cant guarentee) then `for (let key in props) { ret[key] = true; }` would be by far the fastest option.

Comment (rwjblue) [inline packages/ember-metal/lib/mixin.js]: Ahh, I see what you mean, thanks for explaining!

Comment (stefanpenner) [inline packages/ember-metal/lib/mixin.js]: I just thought of it while reading, and thought maybe we should have a quick discussion.

**Comment by marijaselakovic** [comment]
Ok, I squashed all commits into single one. Regarding semantic equivalence, indeed for in+hasOwnProperty check is same as Object.keys. Regarding performance of both cases, if you try to track compiler optimizations in for example V8, you can easily see that V8 disables optimization of for-in loop and reports the following: > ForInStatement is not fast case

In general, the actual performance improvement depends on the input. If .keys function iterates over objects with large number of properties, Object.keys might be performance wise.

**Comment by rwjblue** [comment]
@stefanpenner - I'm going to land this for now (it is a clear improvement), and continue following up with @krisselden about the remaining question.

**Comment by rwjblue** [comment]
Thank you @marijaselakovic!

**Comment by stefanpenner** [comment]
@marijaselakovic I'm confused, how does that differ from my comment. If we know the input is always on the happy path, for in is actually better. My question was largely, can we be sure of that and if not can we make changes to ensure that.

**Comment by marijaselakovic** [comment]
The case I am aware of, when for in and Object.keys perform equally in terms of performance is when you have object without inherited properties and properties that are not dynamically generated, so situation like: var obj={'a':1,'b':2,'c':3}; I am not aware of the case when for in loop is actually better.

**Pull request merged**